File: lib/mailer.ts
import nodemailer from 'nodemailer';

const transporter = nodemailer.createTransport({
  service: 'gmail',
  auth: {
    user: process.env.EMAIL_USER,
    pass: process.env.EMAIL_APP_PASSWORD,
  },
});

export async function sendReminderEmail(to: string, subject: string, html: string): Promise<void> {
  if (!process.env.EMAIL_USER || !process.env.EMAIL_APP_PASSWORD) {
    console.warn(`[mailer] Email not configured — skipping reminder to ${to}`);
    return;
  }
  await transporter.sendMail({ from: `"ILAKA" <${process.env.EMAIL_USER}>`, to, subject, html });
}

export interface TicketEmailData {
  to: string;
  userName: string;
  eventTitle: string;
  eventDate: string;
  ticketId: string;
  rsvpId: string;
  ticketPrice: number | null;
  organizerName: string;
}

function formatINR(paise: number) {
  return new Intl.NumberFormat('en-IN', {
    style: 'currency', currency: 'INR', maximumFractionDigits: 0,
  }).format(paise / 100);
}

export async function sendTicketEmail(data: TicketEmailData) {
  if (!process.env.EMAIL_USER || !process.env.EMAIL_APP_PASSWORD) {
    console.warn('[mailer] EMAIL_USER or EMAIL_APP_PASSWORD not set — skipping email');
    return;
  }

  const ticketUrl = `${process.env.NEXTAUTH_URL ?? 'http://localhost:3000'}/tickets/${data.rsvpId}`;
  const shortId = data.ticketId.slice(0, 8).toUpperCase();
  const amountLine = data.ticketPrice
    ? `<p style="margin:4px 0 0;font-size:13px;color:#b45309;">Amount: <strong>${formatINR(data.ticketPrice)}</strong></p>`
    : '<p style="margin:4px 0 0;font-size:13px;color:#15803d;">Free Entry</p>';

  const html = `
<!DOCTYPE html>
<html lang="en">
<head><meta charset="UTF-8"/><title>Your ILAKA Ticket</title></head>
<body style="margin:0;padding:0;background:#f3f4f6;font-family:system-ui,sans-serif;">
  <table width="100%" cellpadding="0" cellspacing="0" style="background:#f3f4f6;padding:32px 16px;">
    <tr><td align="center">
      <table width="560" cellpadding="0" cellspacing="0" style="background:#ffffff;border-radius:20px;overflow:hidden;box-shadow:0 4px 24px rgba(0,0,0,0.10);">

        <!-- Header -->
        <tr>
          <td style="background:linear-gradient(135deg,#0f766e 0%,#c8663f 100%);padding:32px 36px;">
            <p style="margin:0;font-size:11px;font-weight:700;letter-spacing:0.3em;text-transform:uppercase;color:rgba(255,255,255,0.7);">ILAKA · Event Ticket</p>
            <h1 style="margin:12px 0 6px;font-size:26px;font-weight:700;color:#fff;line-height:1.2;">${data.eventTitle}</h1>
            <p style="margin:0;font-size:13px;color:rgba(255,255,255,0.8);">Hosted by ${data.organizerName}</p>
          </td>
        </tr>

        <!-- Body -->
        <tr>
          <td style="padding:28px 36px;">

            <p style="margin:0 0 20px;font-size:15px;color:#374151;">Hi <strong>${data.userName}</strong>, your spot is confirmed!</p>

            <!-- Ticket details card -->
            <table width="100%" cellpadding="0" cellspacing="0" style="background:#f9fafb;border:1px solid #e5e7eb;border-radius:14px;margin-bottom:20px;">
              <tr>
                <td style="padding:20px 24px;">
                  <table width="100%" cellpadding="0" cellspacing="0">
                    <tr>
                      <td style="padding-bottom:14px;border-bottom:1px solid #e5e7eb;">
                        <p style="margin:0;font-size:10px;font-weight:700;letter-spacing:0.2em;text-transform:uppercase;color:#6b7280;">Ticket No.</p>
                        <p style="margin:4px 0 0;font-size:16px;font-weight:700;font-family:monospace;color:#111;">#${shortId}</p>
                      </td>
                    </tr>
                    <tr>
                      <td style="padding:14px 0;border-bottom:1px solid #e5e7eb;">
                        <p style="margin:0;font-size:10px;font-weight:700;letter-spacing:0.2em;text-transform:uppercase;color:#6b7280;">Attendee</p>
                        <p style="margin:4px 0 0;font-size:15px;font-weight:600;color:#111;">${data.userName}</p>
                      </td>
                    </tr>
                    <tr>
                      <td style="padding:14px 0;border-bottom:1px solid #e5e7eb;">
                        <p style="margin:0;font-size:10px;font-weight:700;letter-spacing:0.2em;text-transform:uppercase;color:#6b7280;">Event</p>
                        <p style="margin:4px 0 0;font-size:15px;font-weight:600;color:#111;">${data.eventTitle}</p>
                        <p style="margin:4px 0 0;font-size:13px;color:#6b7280;">${data.eventDate}</p>
                      </td>
                    </tr>
                    <tr>
                      <td style="padding:14px 0 0;">
                        <p style="margin:0;font-size:10px;font-weight:700;letter-spacing:0.2em;text-transform:uppercase;color:#6b7280;">Payment</p>
                        ${amountLine}
                      </td>
                    </tr>
                  </table>
                </td>
              </tr>
            </table>

            <!-- CTA button -->
            <table width="100%" cellpadding="0" cellspacing="0" style="margin-bottom:20px;">
              <tr>
                <td align="center">
                  <a href="${ticketUrl}" style="display:inline-block;background:linear-gradient(135deg,#0f766e,#c8663f);color:#fff;font-size:14px;font-weight:700;text-decoration:none;padding:14px 32px;border-radius:999px;">View &amp; Download Ticket</a>
                </td>
              </tr>
            </table>

            <p style="margin:0;font-size:12px;color:#9ca3af;text-align:center;">Present your entry QR code at the event entrance.<br/>This ticket is non-transferable.</p>
          </td>
        </tr>

        <!-- Footer -->
        <tr>
          <td style="background:#f9fafb;border-top:1px solid #e5e7eb;padding:16px 36px;text-align:center;">
            <p style="margin:0;font-size:11px;color:#9ca3af;">ILAKA · Community Events Platform · © ${new Date().getFullYear()}</p>
          </td>
        </tr>

      </table>
    </td></tr>
  </table>
</body>
</html>`;

  await transporter.sendMail({
    from: `"ILAKA Events" <${process.env.EMAIL_USER}>`,
    to: data.to,
    subject: `Your ticket for "${data.eventTitle}" — #${shortId}`,
    html,
  });
}

export async function sendVerificationEmail(to: string, verifyUrl: string): Promise<void> {
  if (!process.env.EMAIL_USER || !process.env.EMAIL_APP_PASSWORD) {
    console.log(`[mailer] No email credentials — verify link for ${to}: ${verifyUrl}`);
    return;
  }

  const html = `
<!DOCTYPE html>
<html lang="en">
<head><meta charset="UTF-8"/><title>Verify your ILAKA email</title></head>
<body style="margin:0;padding:0;background:#f3f4f6;font-family:system-ui,sans-serif;">
  <table width="100%" cellpadding="0" cellspacing="0" style="background:#f3f4f6;padding:32px 16px;">
    <tr><td align="center">
      <table width="520" cellpadding="0" cellspacing="0" style="background:#ffffff;border-radius:20px;overflow:hidden;box-shadow:0 4px 24px rgba(0,0,0,0.10);">
        <tr>
          <td style="background:linear-gradient(135deg,#0f766e 0%,#c8663f 100%);padding:32px 36px;">
            <p style="margin:0;font-size:11px;font-weight:700;letter-spacing:0.3em;text-transform:uppercase;color:rgba(255,255,255,0.7);">ILAKA · Email Verification</p>
            <h1 style="margin:12px 0 0;font-size:24px;font-weight:700;color:#fff;">Verify your email</h1>
          </td>
        </tr>
        <tr>
          <td style="padding:32px 36px;">
            <p style="margin:0 0 20px;font-size:15px;color:#374151;">Welcome to ILAKA! Click the button below to verify your email — the link expires in <strong>24 hours</strong>.</p>
            <table width="100%" cellpadding="0" cellspacing="0" style="margin-bottom:24px;">
              <tr>
                <td align="center">
                  <a href="${verifyUrl}" style="display:inline-block;background:linear-gradient(135deg,#0f766e,#c8663f);color:#fff;font-size:14px;font-weight:700;text-decoration:none;padding:14px 32px;border-radius:999px;">Verify my email</a>
                </td>
              </tr>
            </table>
            <p style="margin:0;font-size:13px;color:#6b7280;">If you didn't create an account, you can safely ignore this email.</p>
          </td>
        </tr>
        <tr>
          <td style="background:#f9fafb;border-top:1px solid #e5e7eb;padding:16px 36px;text-align:center;">
            <p style="margin:0;font-size:11px;color:#9ca3af;">ILAKA · Community Events Platform · © ${new Date().getFullYear()}</p>
          </td>
        </tr>
      </table>
    </td></tr>
  </table>
</body>
</html>`;

  await transporter.sendMail({
    from: `"ILAKA" <${process.env.EMAIL_USER}>`,
    to,
    subject: 'Verify your ILAKA email address',
    html,
  });
}

export async function sendApiKeyEmail(to: string, keyPrefix: string): Promise<void> {
  if (!process.env.EMAIL_USER || !process.env.EMAIL_APP_PASSWORD) {
    console.warn(`[mailer] Email not configured — skipping API key email to ${to}`);
    return;
  }

  const html = `
<!DOCTYPE html>
<html lang="en">
<head><meta charset="UTF-8"/><title>Your new ILAKA API key</title></head>
<body style="margin:0;padding:0;background:#f3f4f6;font-family:system-ui,sans-serif;">
  <table width="100%" cellpadding="0" cellspacing="0" style="background:#f3f4f6;padding:32px 16px;">
    <tr><td align="center">
      <table width="520" cellpadding="0" cellspacing="0" style="background:#ffffff;border-radius:20px;overflow:hidden;box-shadow:0 4px 24px rgba(0,0,0,0.10);">
        <tr>
          <td style="background:linear-gradient(135deg,#0f766e 0%,#c8663f 100%);padding:32px 36px;">
            <p style="margin:0;font-size:11px;font-weight:700;letter-spacing:0.3em;text-transform:uppercase;color:rgba(255,255,255,0.7);">ILAKA · API Access</p>
            <h1 style="margin:12px 0 0;font-size:24px;font-weight:700;color:#fff;">New API key created</h1>
          </td>
        </tr>
        <tr>
          <td style="padding:32px 36px;">
            <p style="margin:0 0 16px;font-size:15px;color:#374151;">A new API key starting with <code style="background:#f3f4f6;padding:2px 6px;border-radius:4px;font-family:monospace;">${keyPrefix}…</code> was just created for your account.</p>
            <p style="margin:0 0 20px;font-size:14px;color:#6b7280;">The full key was shown once in your browser and is not stored by ILAKA. If you did not create this key, revoke it immediately from your account settings.</p>
            <p style="margin:0;font-size:13px;color:#9ca3af;">Keep your API key secret — treat it like a password.</p>
          </td>
        </tr>
        <tr>
          <td style="background:#f9fafb;border-top:1px solid #e5e7eb;padding:16px 36px;text-align:center;">
            <p style="margin:0;font-size:11px;color:#9ca3af;">ILAKA · Community Events Platform · © ${new Date().getFullYear()}</p>
          </td>
        </tr>
      </table>
    </td></tr>
  </table>
</body>
</html>`;

  await transporter.sendMail({
    from: `"ILAKA" <${process.env.EMAIL_USER}>`,
    to,
    subject: `New API key created — ${keyPrefix}…`,
    html,
  });
}

export async function sendPasswordResetEmail(to: string, resetUrl: string) {
  if (!process.env.EMAIL_USER || !process.env.EMAIL_APP_PASSWORD) {
    console.log(`[mailer] No email credentials — reset link for ${to}: ${resetUrl}`);
    return;
  }

  const html = `
<!DOCTYPE html>
<html lang="en">
<head><meta charset="UTF-8"/><title>Reset your ILAKA password</title></head>
<body style="margin:0;padding:0;background:#f3f4f6;font-family:system-ui,sans-serif;">
  <table width="100%" cellpadding="0" cellspacing="0" style="background:#f3f4f6;padding:32px 16px;">
    <tr><td align="center">
      <table width="520" cellpadding="0" cellspacing="0" style="background:#ffffff;border-radius:20px;overflow:hidden;box-shadow:0 4px 24px rgba(0,0,0,0.10);">
        <tr>
          <td style="background:linear-gradient(135deg,#0f766e 0%,#c8663f 100%);padding:32px 36px;">
            <p style="margin:0;font-size:11px;font-weight:700;letter-spacing:0.3em;text-transform:uppercase;color:rgba(255,255,255,0.7);">ILAKA · Account Recovery</p>
            <h1 style="margin:12px 0 0;font-size:24px;font-weight:700;color:#fff;">Reset your password</h1>
          </td>
        </tr>
        <tr>
          <td style="padding:32px 36px;">
            <p style="margin:0 0 20px;font-size:15px;color:#374151;">We received a request to reset your ILAKA password. Click the button below — the link expires in <strong>1 hour</strong>.</p>
            <table width="100%" cellpadding="0" cellspacing="0" style="margin-bottom:24px;">
              <tr>
                <td align="center">
                  <a href="${resetUrl}" style="display:inline-block;background:linear-gradient(135deg,#0f766e,#c8663f);color:#fff;font-size:14px;font-weight:700;text-decoration:none;padding:14px 32px;border-radius:999px;">Set new password</a>
                </td>
              </tr>
            </table>
            <p style="margin:0;font-size:13px;color:#6b7280;">If you didn't request this, you can safely ignore this email — your password won't change.</p>
          </td>
        </tr>
        <tr>
          <td style="background:#f9fafb;border-top:1px solid #e5e7eb;padding:16px 36px;text-align:center;">
            <p style="margin:0;font-size:11px;color:#9ca3af;">ILAKA · Community Events Platform · © ${new Date().getFullYear()}</p>
          </td>
        </tr>
      </table>
    </td></tr>
  </table>
</body>
</html>`;

  await transporter.sendMail({
    from: `"ILAKA" <${process.env.EMAIL_USER}>`,
    to,
    subject: 'Reset your ILAKA password',
    html,
  });
}

File: backend/app/integrations/mailer.py
"""
Async email system via Gmail SMTP.
Silently skips if EMAIL_USER / EMAIL_APP_PASSWORD are not configured.
"""
import logging
import smtplib
from dataclasses import dataclass
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText

import aiosmtplib

from app.core.config import settings

logger = logging.getLogger(__name__)


@dataclass
class TicketEmailData:
    to: str
    ticket_id: str
    rsvp_id: str
    event_title: str
    event_start: str
    event_location: str
    amount: int | None
    ticket_url: str


async def _send(to: str, subject: str, html: str) -> bool:
    if not settings.email_configured:
        logger.debug("Email not configured, skipping send to %s", to)
        return False

    msg = MIMEMultipart("alternative")
    msg["Subject"] = subject
    msg["From"] = f"{settings.EMAIL_FROM_NAME} <{settings.EMAIL_USER}>"
    msg["To"] = to
    msg.attach(MIMEText(html, "html"))

    try:
        await aiosmtplib.send(
            msg,
            hostname=settings.SMTP_HOST,
            port=settings.SMTP_PORT,
            start_tls=True,
            username=settings.EMAIL_USER,
            password=settings.EMAIL_APP_PASSWORD,
        )
        logger.info("Email sent to %s: %s", to, subject)
        return True
    except Exception as exc:
        logger.error("Failed to send email to %s: %s", to, exc)
        return False


async def send_ticket_email(data: TicketEmailData) -> bool:
    price_str = (
        f"₹{data.amount / 100:.2f}" if data.amount else "Free"
    )
    html = f"""
    <div style="font-family:sans-serif;max-width:600px;margin:auto">
      <h2 style="color:#6d28d9">You're in! 🎉</h2>
      <p>Here's your ticket for <strong>{data.event_title}</strong></p>
      <table style="width:100%;border-collapse:collapse;margin:16px 0">
        <tr><td style="padding:8px;font-weight:bold">Ticket #</td>
            <td style="padding:8px">{data.ticket_id}</td></tr>
        <tr style="background:#f5f3ff">
            <td style="padding:8px;font-weight:bold">Date</td>
            <td style="padding:8px">{data.event_start}</td></tr>
        <tr><td style="padding:8px;font-weight:bold">Location</td>
            <td style="padding:8px">{data.event_location}</td></tr>
        <tr style="background:#f5f3ff">
            <td style="padding:8px;font-weight:bold">Amount</td>
            <td style="padding:8px">{price_str}</td></tr>
      </table>
      <a href="{data.ticket_url}"
         style="display:inline-block;padding:12px 24px;background:#6d28d9;
                color:white;text-decoration:none;border-radius:6px">
        View &amp; Download Ticket
      </a>
      <p style="color:#888;font-size:12px;margin-top:24px">
        This is an automated message from Illaka. Please do not reply.
      </p>
    </div>
    """
    return await _send(data.to, f"Your ticket for {data.event_title}", html)


async def send_reminder_email(
    to: str,
    subject: str,
    event_title: str,
    event_start: str,
    is_online: bool,
    online_link: str | None = None,
) -> bool:
    location_block = (
        f'<p>Join online: <a href="{online_link}">{online_link}</a></p>'
        if is_online and online_link
        else "<p>See the event page for location details.</p>"
    )
    html = f"""
    <div style="font-family:sans-serif;max-width:600px;margin:auto">
      <h2 style="color:#6d28d9">Reminder: {event_title}</h2>
      <p>Your event starts on <strong>{event_start}</strong></p>
      {location_block}
      <p style="color:#888;font-size:12px;margin-top:24px">
        You received this because you RSVPed on Illaka.
      </p>
    </div>
    """
    return await _send(to, subject, html)


def send_reminder_email_sync(
    to: str,
    user_name: str,
    event_title: str,
    event_date: str,
    event_type: str,
    online_link: str | None = None,
) -> None:
    """Synchronous reminder email for Celery workers (uses smtplib, not aiosmtplib)."""
    if not settings.email_configured:
        return

    location_block = (
        f'<p>Join online: <a href="{online_link}">{online_link}</a></p>'
        if event_type == "ONLINE" and online_link
        else "<p>See the event page for location details.</p>"
    )
    html = f"""
    <div style="font-family:sans-serif;max-width:600px;margin:auto">
      <h2 style="color:#6d28d9">Reminder: {event_title}</h2>
      <p>Hi {user_name}, your event starts on <strong>{event_date}</strong></p>
      {location_block}
      <p style="color:#888;font-size:12px;margin-top:24px">
        You received this because you RSVPed on Illaka.
      </p>
    </div>
    """
    msg = MIMEMultipart("alternative")
    msg["Subject"] = f"Reminder: {event_title}"
    msg["From"] = f"{settings.EMAIL_FROM_NAME} <{settings.EMAIL_USER}>"
    msg["To"] = to
    msg.attach(MIMEText(html, "html"))

    with smtplib.SMTP(settings.SMTP_HOST, settings.SMTP_PORT) as conn:
        conn.ehlo()
        conn.starttls()
        conn.login(settings.EMAIL_USER, settings.EMAIL_APP_PASSWORD)
        conn.sendmail(settings.EMAIL_USER, to, msg.as_string())


async def send_password_reset_email(to: str, reset_url: str) -> bool:
    html = f"""
    <div style="font-family:sans-serif;max-width:600px;margin:auto">
      <h2 style="color:#6d28d9">Reset your password</h2>
      <p>Click the button below to set a new password. This link expires in 1 hour.</p>
      <a href="{reset_url}"
         style="display:inline-block;padding:12px 24px;background:#6d28d9;
                color:white;text-decoration:none;border-radius:6px">
        Set new password
      </a>
      <p style="color:#888;font-size:12px;margin-top:24px">
        If you did not request this, ignore this email.
      </p>
    </div>
    """
    return await _send(to, "Reset your Illaka password", html)
